### PR TITLE
New module: moran-index.

### PR DESCRIPTION
**Pull request by zhuang-hao-ming**
Add a new module moran-index and a assistant new module distance-weight.

moran-index can be used to measure the patterns of attribute values associated with features.
A positive one means similar values tend to occur near each other, while a negative one means
high and low values are interspersed.

distance-weight can make a distance-weight-matrix from feature collection based on plane coordinate.

## polygon cluster
moran_polygon
```json
{
    "moranI": 0.1485081274747776,
    "expectMoranI": -0.020833333333333332,
    "stdNorm": 0.02374513825431575,
    "zNorm": 7.131626651082253
}
```
## point cluster
moran_point
```json
{
    "moranI": 0.15665417693293948,
    "expectMoranI": -0.020833333333333332,
    "stdNorm": 0.022208244679327364,
    "zNorm": 7.991964823383264
}
```

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Have read [How To Contribute](https://github.com/Turfjs/turf/blob/master/CONTRIBUTING.md#how-to-contribute).
- [x] Run `npm test` at the sub modules where changes have occurred.
- [x] Run `npm run lint` to ensure code style at the turf module level.

Submitting a new TurfJS Module.

- [x] Overview description of proposed module.
- [x] Include JSDocs with a basic example.
- [x] Execute `./scripts/generate-readmes` to create `README.md`.
- [x] Add yourself to **contributors** in `package.json` using "Full Name <@GitHub Username>".

### CODE REVIEW

File: packages/turf-moran-index/index.ts [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+ *
+ * @param {FeatureCollection<any>} fc
+ * @param {Object} options
+ * @param {string} options.inputField the property name
+ * @param {number} [options.threshold] the distance threshold
+ * @param {number} [options.p] the Minkowski p-norm distance parameter
+ * @param {boolean} [options.binary] whether transfrom the distance to binary
+ * @param {number} [options.alpha] the distance decay parameter
+ * @param {boolean} [options.standardization] wheter row standardization the distance
+ * @returns
```
Comment (DenisCarriere) [inline packages/turf-moran-index/index.ts]: Should return an `{Object}` or even better you can define your own JSDocs `@typedef`.

http://usejsdoc.org/tags-typedef.html

**Example**

```js
/**
 * MoranIndex - DESCRIPTION
 * @typedef {Object} MoranIndex
 * @property {number} moranI - DESCRIPTION
 * @property {number} expectMoranI - DESCRIPTION
 * @property {number} stdNorm - DESCRIPTION
 * @property {number} zNorm - DESCRIPTION
 */
```

Returns would look like this:

```js
/**
 * @returns {MoranIndex}
```

File: packages/turf-moran-index/index.ts [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+ * 1. [Moran's I](https://en.wikipedia.org/wiki/Moran%27s_I)
+ * 
+ * 2. [pysal](http://pysal.readthedocs.io/en/latest/index.html)
+ * 
+ * 3. Andy Mitchell, The ESRI Guide to GIS Analysis Volume 2: Spatial Measurements & Statistics.
+ *
+ * @param {FeatureCollection<any>} fc
+ * @param {Object} options
+ * @param {string} options.inputField the property name
+ * @param {number} [options.threshold] the distance threshold
```
Comment (DenisCarriere) [inline packages/turf-moran-index/index.ts]: Add default values:

```js
/**
* @param {number} [options.threshold= 100000] the distance threshold
```

File: packages/turf-moran-index/index.ts [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+ * @param {number} [options.p] the Minkowski p-norm distance parameter
```
Comment (DenisCarriere) [inline packages/turf-moran-index/index.ts]: Add default values for all `options` fields

File: packages/turf-distance-weight/index.ts [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+  return Math.pow((Math.pow(xDiff, p) + Math.pow(yDiff, p)), 1 / p);
+}
+
+/**
+ *
+ *
+ * @name distanceWeight
+ * @param {FeatureCollection<any>} fc FeatureCollection.
+ * @param {Object} [options] option object.
+ * @param {number} [options.threshold] If the distance between neighbor and target features is greater than threshold,
```
Comment (DenisCarriere) [inline packages/turf-distance-weight/index.ts]: Add `default` values to all `options` fields.

ex:

```js
/**
 * @param {number} [options.threshold= 10000]
```

**Comment by rowanwins** [comment]
Thanks @zhuang-hao-ming 

A couple of minor comments

1. For the moran-index I'd also suggest adding either in the description or in the `options.inputField` that this is designed for working with numeric properties, might be kinda obvious but something worth stating :) 
````
@param {string} options.inputField the property name/key - must contain numeric values
`````

2. For the return object of the moran index I'd suggest renaming `moranI` to `moranIndex` and `expectMoranI` to `expectMoranIndex` just so it's blindingly clear. 

Other than those minor quibbles looks really interesting!

**Comment by zhuang-hao-ming** [comment]
@DenisCarriere 
Thanks for your reviews.
I have changed the JSDocs as your suggestions.
As for that distance-weight module, because I will use it in many places, 
such as other spatial autocorrelation indices, so i let it be a module alone.
And we will add more spatial-weight-matrix related modules, such as `contiguity weight`, `knn weight`, `kernal weight`.With these modules. We can make a lot of spatial statistics modules.

**Comment by zhuang-hao-ming** [comment]
Hi @rowanwins.
Thanks for your reviews.
I  agree with your opinions, and I have revised the code as you suggest.

**Comment by DenisCarriere** [comment]
> We can make a lot of spatial statistics modules.😃

Amazing stuff! Looking forward to it!

**Comment by DenisCarriere** [comment]
@zhuang-hao-ming Published on `npm` as `v6.0.0` release.

https://unpkg.com/@turf/moran-index@6.0.0/

**Comment by DenisCarriere** [comment]
And published `@turf/distance-weight` as `v6.0` release 🎉 

https://unpkg.com/@turf/distance-weight@6.0.0/

**Pull request merged**